internals-js/src/federation.ts

```diff
@@ -336,6 +344,337 @@ function fieldSetTargetDescription(directive: Directive<any, {fields: any}>): st
   return `${targetKind} "${directive.parent?.coordinate}"`;
 }
 
+export function parseContext(input: string) {
+  const regex = /^(?:[\n\r\t ,]|#[^\n\r]*(?![^\n\r]))*\$(?:[\n\r\t ,]|#[^\n\r]*(?![^\n\r]))*([A-Za-z_]\w*(?!\w))([\s\S]*)$/;
+  const match = input.match(regex);
+  if (!match) {
+    return { context: undefined, selection: undefined };
+  }
+
+  const [, context, selection] = match;
+  return {
+    context,
+    selection,
+  };
+}
+
+const wrapResolvedType = ({
+  originalType,
+  resolvedType,
+}: {
+  originalType: OutputType,
+  resolvedType: InputType,
+}): InputType | undefined => {
+  const stack = [];
+  let unwrappedType: NamedType | WrapperType = originalType;
+  while(unwrappedType.kind === 'NonNullType' || unwrappedType.kind === 'ListType') {
+    stack.push(unwrappedType.kind);
+    unwrappedType = unwrappedType.baseType();
+  }
+
+  let type: NamedType | WrapperType = resolvedType;
+  while(stack.length > 0) {
+    const kind = stack.pop();
+    if (kind === 'NonNullType' && type.kind !== 'NonNullType') {
+      type = new NonNullType(type);
+    } else if (kind === 'ListType') {
+      type = new ListType(type);
+    }
```

Comment:
I've realized that similar to the issue where other subgraphs introducing a member to an abstract type can "introduce a nullability", other subgraphs can similarly change the return type of a field from non-nullable to nullable (e.g. if a `@fromContext` in Subgraph A uses a field `foo` with return type `String!`, and Subgraph B declares field `foo` with return type `String`, `foo`s return type will be `String` in the supergraph schema).

For now, similar to the case with new member types, I think the simplest solution is to assume it's always nullable. The effect then, is that these lines will change to just being
```typescript
if (kind === 'ListType') {
  type = new ListType(type);
}
```
I.e. we won't add a `NonNullType` wrapper even if `kind` is `NonNullType`, and it'll be skipped.

In a previous comment, I mentioned that in the future, we could precompute the possible runtime types in the supergraph schema to fix nullability introduced through additions in abstract type membership. However, with this additional way to "introduce nullability", I think the approach in the future should change to:
1. In `onValidation()`, we compute the "resolved type" completely unaware of any nullabilities introduced by other subgraphs, and ensure this is compatible with the argument's type. This will catch errors that are subgraph specific.
2. As a post-merge validation, we parse contexts against the supergraph schema, and recompute the "resolved type", and ensure this is compatible with the argument's type. This will catch errors caused by other subgraphs introducing nullability. (Note that post-merge, we don't have to re-check other context constraints, e.g. that there's only field executed.)